### PR TITLE
chore: add jenkinsfile

### PR DESCRIPTION
**Pull request by pchandra19**
Signed-off-by: Prateek Chandra <Prateek.Chandra@datacore.com>

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    run_tests = false
+}
+
+// Only schedule regular builds on main branches, so we don't need to guard against it
+String cron_schedule = mainBranches() ? "0 2 * * *" : ""
+
+pipeline {
+  agent none
+  options {
+    timeout(time: 1, unit: 'HOURS')
```
Comment (tiagolobocastro) [inline Jenkinsfile]: Maybe change to 2 hours.
I've already made this change in control-plane as now that we have loads of nightly runs the agents get busy and it takes longer to get a free agent to run on.
We should probably consider adding more agents to Jenkins.

File: Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/env groovy
+
+def mainBranches() {
+    return BRANCH_NAME == "develop" || BRANCH_NAME.startsWith("release/");
+}
+
+// TODO: Use multiple choices
+run_linter = true
+rust_test = true
```
Comment (tiagolobocastro) [inline Jenkinsfile]: not needed

File: Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+      }
+      parallel {
+        stage('chart publish test') {
+          when {
+            expression { helm_test == true }
+          }
+          agent { label 'nixos-mayastor' }
+          steps {
+            sh 'printenv'
+            sh 'nix-shell --pure --run "./scripts/helm/test-publish-chart-yaml.sh" ./scripts/helm/shell.nix'
```
Comment (tiagolobocastro) [inline Jenkinsfile]: we should move the shell.nix to the root now

File: Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+      steps {
+        sh 'printenv'
+      }
+    }
+    stage('test') {
+      when {
+        beforeAgent true
+        not {
+          anyOf {
+            branch 'master'
```
Comment (tiagolobocastro) [inline Jenkinsfile]: don't have a master concept anymore

File: Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    }
+    stage('linter') {
+      agent { label 'nixos-mayastor' }
+      when {
+        beforeAgent true
+        not {
+          anyOf {
+            branch 'master'
+            branch 'release/*'
+            expression { run_linter == false }
```
Comment (tiagolobocastro) [inline Jenkinsfile]: we don't have a linter here, but I guess we could use a linter to lint values.yaml in the future?

**Review by tiagolobocastro — APPROVED** [review]

File: Jenkinsfile (hunk 1 above)
Comment (pchandra19) [reply Jenkinsfile]: Done.

File: Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+          contextSource: [
+            $class: 'ManuallyEnteredCommitContextSource',
+            context: 'continuous-integration/jenkins/branch'
+          ],
+          statusMessage: [ content: 'Pipeline started' ]
+        ])
+      }
+    }
+    stage('test') {
+      parallel {
```
Comment (blaisedias) [inline Jenkinsfile]: There is only one stage - so why the parallel block?

**Comment by pchandra19** [comment]
bors try

**Review by blaisedias — APPROVED** [review]

**Pull request merged**